README.md
# appear-event

Element appear event.

[![Build Status][travis-image]][travis-url]
[![NPM Version][npm-version-image]][npm-url]
[![NPM Downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![Standard Version][standard-version-image]][standard-version-url]
[![Codecov][codecov-image]][codecov-url]

# Install

`yarn add appear-event` or `npm i appear-event --save`

# Usage

## `observe` and `unobserve`

```js
const { observe, unobserve } = require('appear-event');

const el = getElementSomehow();
const eventListenerOptions = getEventListenerOptions();
const intersectionObserverOptions = getIntersectionObserverOptions();

function onAppear(appearEvent) {
  // element appeared
}

function onDisappear(disappearEvent) {
  // element disappeared
}

observe(el, intersectionObserverOptions); // watch for appear and disappear event
el.addEventListener('appear', onAppear, eventListenerOptions);
el.addEventListener('disappear', onDisappear, eventListenerOptions);
el.removeEventListener('appear', onAppear, eventListenerOptions);
el.removeEventListener('disappear', onDisappear, eventListenerOptions);
unobserve(el, intersectionObserverOptions); // unwatch for appear and disappear event
```

### `eventListenerOptions`

[addEventListener options](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Parameters)

### `intersectionObserverOptions`

[IntersectionObserver Properties](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver#Properties)

### `appearEvent` and `disappearEvent`

[CustomEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent)

#### `detail`

`appearEvent.detail` and `disappearEvent.detail` is an [IntersectionObserveEntry](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry)

## event-listeners

```js
const { addAppearListener, addDisappearListener, removeAppearListener, removeDisappearListener } = require('appear-event/lib/event-listener');

const el = getElementSomehow();
const eventListenerOptions = getEventListenerOptions();
const intersectionObserverOptions = getIntersectionObserverOptions();

function onAppear(appearEvent) {
  // element appeared
}

function onDisappear(disappearEvent) {
  // element disappeared
}

addAppearListener(el, onAppear, { ...eventListenerOptions, ...intersectionObserverOptions });
addDisappearListener(el, onDisappear, { ...eventListenerOptions, ...intersectionObserverOptions });
removeAppearListener(el, onAppear, { ...eventListenerOptions, ...intersectionObserverOptions });
removeDisappearListener(el, onDisappear, { ...eventListenerOptions, ...intersectionObserverOptions });
```

## In React

See [demo](./demo/src/App.js)
Also see [jsx-native-events](https://github.com/calebdwilliams/jsx-native-events)

## FAQ

**Q:** Does zero area elements appears when they are in the screen?

**A:** Yes. And if you'd like to tell if the element is a zero area element, use `event.detail.boundingClientRect` to get the element dimensions.

#

Project created by [create-n](https://github.com/vivaxy/create-n).

[travis-image]: https://img.shields.io/travis/vivaxy/appear-event.svg?style=flat-square
[travis-url]: https://travis-ci.org/vivaxy/appear-event
[npm-version-image]: https://img.shields.io/npm/v/appear-event.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/appear-event
[npm-downloads-image]: https://img.shields.io/npm/dt/appear-event.svg?style=flat-square
[license-image]: https://img.shields.io/npm/l/appear-event.svg?style=flat-square
[license-url]: LICENSE
[standard-version-image]: https://img.shields.io/badge/release-standard%20version-brightgreen.svg?style=flat-square
[standard-version-url]: https://github.com/conventional-changelog/standard-version
[codecov-image]: https://img.shields.io/codecov/c/github/vivaxy/appear-event.svg?style=flat-square
[codecov-url]: https://codecov.io/gh/vivaxy/appear-event


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.5](https://github.com/vivaxy/appear-event/compare/v0.3.4...v0.3.5) (2020-05-25)


### Bug Fixes

* :bug: fix @medv/finder@2.0.0 contains es2015 codes ([4e41f92](https://github.com/vivaxy/appear-event/commit/4e41f927748001130c1be34ad642c91540da66ab))

### [0.3.4](https://github.com/vivaxy/appear-event/compare/v0.3.3...v0.3.4) (2020-05-25)


### Bug Fixes

* :bug: upgrade finder ([bae65ae](https://github.com/vivaxy/appear-event/commit/bae65ae1d0553ac90974210459c5b0c1fa0e34ea))
* **deps:** update dependency @medv/finder to v2 ([2a9e354](https://github.com/vivaxy/appear-event/commit/2a9e3546bb7352db11d5d3e25f84b4ae2b3863d9))

### [0.3.3](https://github.com/vivaxy/appear-event/compare/v0.3.2...v0.3.3) (2020-05-20)


### Bug Fixes

* fix when page loads, disappear may happens ([46eff4e](https://github.com/vivaxy/appear-event/commit/46eff4e8dbc62f34d6f3cc506ee8a06bbc73ab3f))
* **deps:** update dependency intersection-observer to ^0.10.0 ([c56834d](https://github.com/vivaxy/appear-event/commit/c56834d2d4928a11c9b3f608296940d41db24ac6))
* **deps:** update dependency react-scripts to v3.4.1 ([67811d4](https://github.com/vivaxy/appear-event/commit/67811d42b06274f850d18416e794fc8c5a4e4de8))

### [0.3.2](https://github.com/vivaxy/appear-event/compare/v0.3.1...v0.3.2) (2019-11-27)


### Bug Fixes

* **compatiable:** :bug: fix Array.prototype.includes fails on android 4 ([c9f3b95](https://github.com/vivaxy/appear-event/commit/c9f3b95))



### [0.3.1](https://github.com/vivaxy/appear-event/compare/v0.3.0...v0.3.1) (2019-07-30)


### Bug Fixes

* **es:** :bug: exporting es5 ([db12912](https://github.com/vivaxy/appear-event/commit/db12912))



## [0.3.0](https://github.com/vivaxy/appear-event/compare/v0.2.0...v0.3.0) (2019-07-29)


### Features

* **event listeners:** :sparkles: support usage for event-listeners ([9f36a1a](https://github.com/vivaxy/appear-event/commit/9f36a1a))



## [0.2.0](https://github.com/vivaxy/appear-event/compare/v0.1.1...v0.2.0) (2019-07-27)


### Features

* **event listeners:** :fire: remove addAppearEventListener, addDisappea ([9bae199](https://github.com/vivaxy/appear-event/commit/9bae199))


### Tests

* **test:** :white_check_mark: ([b045285](https://github.com/vivaxy/appear-event/commit/b045285))



### [0.1.1](https://github.com/vivaxy/appear-event/compare/v0.1.0...v0.1.1) (2019-07-27)


### Tests

* **coverage:** :white_check_mark: code coverage from v8 as 1:many is no ([c7c88d0](https://github.com/vivaxy/appear-event/commit/c7c88d0))
* **coverage:** :white_check_mark: fix coverage ([df53e73](https://github.com/vivaxy/appear-event/commit/df53e73))
* **test:** :construction_worker: update coverage report ([0660282](https://github.com/vivaxy/appear-event/commit/0660282))
* **test:** :white_check_mark: add test cases ([703de92](https://github.com/vivaxy/appear-event/commit/703de92))
* **test:** :white_check_mark: fix test ([baac349](https://github.com/vivaxy/appear-event/commit/baac349))
* **test:** :white_check_mark: fix test coverage ([3733e56](https://github.com/vivaxy/appear-event/commit/3733e56))
* **test:** :white_check_mark: update test ([792ec52](https://github.com/vivaxy/appear-event/commit/792ec52))



## 0.1.0 (2019-07-26)


### Features

* **appear:** :tada: implementation ([220e8de](https://github.com/vivaxy/appear-event/commit/220e8de))


src/__tests__/index.test.js
/**
 * @since 20180911 17:16
 * @author vivaxy
 */
async function sleep(timeout = 300) {
  return new Promise(function(resolve) {
    setTimeout(resolve, timeout);
  });
}

beforeAll(async () => {
  await page.coverage.startJSCoverage();
  await page.emulate({
    viewport: {
      width: 375,
      height: 667,
      deviceScaleFactor: 2,
      isMobile: true,
      hasTouch: true,
      isLandscape: false,
    },
    userAgent:
      'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1',
  });
  await page.goto('http://localhost:4444');
});

afterAll(async function() {
  const jsCoverage = await page.coverage.stopJSCoverage();
  const v8toIstanbul = require('v8-to-istanbul');
  const fse = require('fs-extra');
  const converter = v8toIstanbul('./src/__tests__/helpers/index.bundle.js');
  await converter.load();
  converter.applyCoverage(jsCoverage);
  await fse.outputFile(
    '.nyc_output/out.json',
    JSON.stringify(converter.toIstanbul()),
  );
});

it('should_appear', async () => {
  await sleep();
  await expect(page.title()).resolves.toMatch('should_appear');
});

it('should_disappear', async () => {
  await page.click('#should_disappear');
  await sleep();
  await expect(page.title()).resolves.toMatch('should_disappear');
});

it('should_appear_again', async () => {
  await page.click('#should_appear_again');
  await sleep();
  await expect(page.title()).resolves.toMatch('should_appear_again');
});


src/__tests__/index.test.ts
/**
 * @since 20180911 17:16
 * @author vivaxy
 */
beforeAll(async () => {
  await page.emulate({
    viewport: {
      width: 375,
      height: 667,
      deviceScaleFactor: 2,
      isMobile: true,
      hasTouch: true,
      isLandscape: false,
    },
    userAgent:
      'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1',
  });
  await page.goto('http://localhost:4444');
});

it('should_appear', async () => {
  await expect(page.title()).resolves.toMatch('should_appear');
});

it('should_disappear', async () => {
  await page.click('#should_disappear');
  await expect(page.title()).resolves.toMatch('should_disappear');
});

// it('should_appear_again', async () => {
//   await page.click('#should_appear_again');
//   await expect(page.title()).resolves.toMatch('should_appear_again');
// });


src/__tests__/helpers/index.js
/**
 * @since 2019-07-26 20:02
 * @author vivaxy
 */

import * as appearEvent from '../../index.ts';

let testSeq = 0;

const $items = document.querySelectorAll('.item');
$items[0].addEventListener('appear', function () {
  testSeq++;
  switch (testSeq) {
    case 1:
      document.title = 'should_appear';
      break;
    case 3:
      document.title = 'should_appear_again';
      break;
    default:
      throw new Error('Unhandled testSeq: ' + testSeq);
  }
});
$items[0].addEventListener('disappear', function () {
  testSeq++;
  switch (testSeq) {
    case 2:
      document.title = 'should_disappear';
      break;
    case 4:
      // nothing
      break;
    default:
      throw new Error('Unhandled testSeq: ' + testSeq);
  }
});
appearEvent.observe($items[0]);

// const $zeroHeight = document.getElementById('zero-height');
// $zeroHeight.addEventListener('appear', function () {
//   throw new Error('Should never appear');
// });
// appearEvent.observe($zeroHeight);

// const $outOfBounds = document.getElementById('out-of-bounds');
// $outOfBounds.addEventListener('appear', function () {
//   throw new Error('Should never appear');
// });
// appearEvent.observe($outOfBounds);


src/index.ts
import 'intersection-observer';
import finder from '@medv/finder';

export const observingCollection: {
  [serializedIntersectionObserverOptions: string]: {
    serializedOptions: string;
    intersectionObserver: IntersectionObserver;
    elements: Element[];
  };
} = {};

function serializeIntersectionObserverOptions(
  options?: IntersectionObserverInit,
) {
  if (!options) {
    return '';
  }
  const serializer = {
    root(root: undefined | Element | null) {
      if (!root) {
        return '';
      }
      return finder(root);
    },
    rootMargin(rootMargin: undefined | string) {
      return rootMargin || '';
    },
    threshold(threshold: undefined | number | number[]) {
      if (threshold === undefined) {
        return '';
      }
      return String(threshold);
    },
  };
  return Object.keys(serializer)
    .map(function (key) {
      // @ts-ignore TODO fix types
      return serializer[key](options[key]);
    })
    .join('; ');
}

function intersectionObserverCallback(entries: IntersectionObserverEntry[]) {
  entries.forEach(function (entry) {
    const el = entry.target as Element & { appeared?: boolean };
    if (entry.isIntersecting && !el.appeared) {
      el.appeared = true;
      const appearEvent = new CustomEvent<IntersectionObserverEntry>('appear', {
        detail: entry,
      });
      el.dispatchEvent(appearEvent);
    } else if (el.appeared) {
      delete el.appeared;
      const disappearEvent = new CustomEvent<IntersectionObserverEntry>(
        'disappear',
        { detail: entry },
      );
      el.dispatchEvent(disappearEvent);
    }
  });
}

function getObserving(options?: IntersectionObserverInit) {
  const serialized = serializeIntersectionObserverOptions(options);
  const observing = observingCollection[serialized];
  if (observing) {
    return observing;
  }
  const intersectionObserver = new IntersectionObserver(
    intersectionObserverCallback,
    options,
  );
  const newObserving = {
    serializedOptions: serialized,
    intersectionObserver,
    elements: [],
  };
  observingCollection[serialized] = newObserving;
  return newObserving;
}

export function observe(el: Element, options?: IntersectionObserverInit): void {
  if (!el) {
    return;
  }
  const observing = getObserving(options);
  if (observing.elements.indexOf(el) === -1) {
    observing.intersectionObserver.observe(el);
    observing.elements.push(el);
  }
}

export function unobserve(
  el: Element,
  options?: IntersectionObserverInit,
): void {
  if (!el) {
    return;
  }
  // strict observing retrieving
  const observing = getObserving(options);
  const { elements, intersectionObserver, serializedOptions } = observing;
  if (elements.indexOf(el) !== -1) {
    intersectionObserver.unobserve(el);
    elements.splice(elements.indexOf(el), 1);
    if (!elements.length) {
      intersectionObserver.disconnect();
      delete observingCollection[serializedOptions];
    }
  }
}


demo/src/App.js
/** @jsx nativeEvents */
import React, { Component, createRef } from 'react';
import nativeEvents from 'jsx-native-events';
import { observe, unobserve } from './lib';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.ref = createRef();
  }

  componentDidMount() {
    observe(this.ref.current);
  }

  componentWillUnmount() {
    unobserve(this.ref.current);
  }

  handleAppear(e) {
    console.log('appear', e);
  }

  handleDisappear(e) {
    console.log('disappear', e);
  }

  render() {
    return (
      <div>
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div
          className="item"
          onEventAppear={this.handleAppear}
          onEventDisappear={this.handleDisappear}
          ref={this.ref}
        >
          appear and disappear
        </div>
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
        <div className="placeholder" />
      </div>
    );
  }
}


src/event-listener.ts
/**
 * @since 2019-07-29 13:06
 * @author vivaxy
 */
import { observe, unobserve } from './index';

type AppearEvent = CustomEvent<IntersectionObserverEntry>;

interface OnEvent {
  (e: AppearEvent): any;
}

type Options = boolean | (AddEventListenerOptions & IntersectionObserverInit);

export const EVENT_LISTENERS = Symbol('appear-event/event-listeners');

export enum EVENT_TYPE {
  APPEAR = 'appear',
  DISAPPEAR = 'disappear',
}

interface ElementWithEventListeners extends Element {
  [EVENT_LISTENERS]: {
    type: EVENT_TYPE;
    handler: OnEvent;
    options?: Options;
  }[];
}

function prepareOptions(
  options?: Options,
): [
  undefined | IntersectionObserverInit,
  undefined | boolean | AddEventListenerOptions,
] {
  const intersectionObserverOptions =
    typeof options === 'undefined'
      ? undefined
      : typeof options === 'boolean'
      ? undefined
      : {
          root: options.root,
          rootMargin: options.rootMargin,
          threshold: options.threshold,
        };
  const eventListenerOptions =
    typeof options === 'undefined'
      ? undefined
      : typeof options === 'boolean'
      ? options
      : {
          once: options.once,
          passive: options.passive,
          capture: options.capture,
        };
  return [intersectionObserverOptions, eventListenerOptions];
}

function toEventListenerElement(el: Element): ElementWithEventListeners {
  (<ElementWithEventListeners>el)[EVENT_LISTENERS] =
    (<ElementWithEventListeners>el)[EVENT_LISTENERS] || [];
  return <ElementWithEventListeners>el;
}

export function addEventListener(
  ele: Element,
  type: EVENT_TYPE,
  onEvent: OnEvent,
  options?: Options,
) {
  const el = toEventListenerElement(ele);
  const [intersectionObserverOptions, eventListenerOptions] = prepareOptions(
    options,
  );
  observe(el, intersectionObserverOptions);
  el.addEventListener(type, onEvent as EventListener, eventListenerOptions);
  el[EVENT_LISTENERS].push({
    type,
    handler: onEvent,
    options,
  });
}

export function addAppearListener(
  el: Element,
  onAppear: OnEvent,
  options?: Options,
) {
  addEventListener(el, EVENT_TYPE.APPEAR, onAppear, options);
}

export function addDisappearListener(
  el: Element,
  onDisappear: OnEvent,
  options?: Options,
) {
  addEventListener(el, EVENT_TYPE.DISAPPEAR, onDisappear, options);
}

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener
 */
function optionsMatches(options1?: Options, options2?: Options): boolean {
  if (options1 === options2) {
    return true;
  }
  return !!(
    typeof options1 === 'object' &&
    typeof options2 === 'object' &&
    options1.capture === options2.capture
  );
}

export function removeEventListener(
  ele: Element,
  type: EVENT_TYPE,
  onEvent: OnEvent,
  options?: Options,
) {
  const el = toEventListenerElement(ele);
  const [intersectionObserverOptions, eventListenerOptions] = prepareOptions(
    options,
  );
  const listenerIndex = el[EVENT_LISTENERS].findIndex(function(listener) {
    return (
      listener.type === type &&
      listener.handler === onEvent &&
      optionsMatches(listener.options, eventListenerOptions)
    );
  });
  if (listenerIndex !== -1) {
    el.removeEventListener(
      type,
      onEvent as EventListener,
      eventListenerOptions,
    );
    el[EVENT_LISTENERS].splice(listenerIndex, 1);
  }
  if (el[EVENT_LISTENERS].length === 0) {
    delete el[EVENT_LISTENERS];
    unobserve(el, intersectionObserverOptions);
  }
}

export function removeAppearListener(
  el: Element,
  onAppear: OnEvent,
  options?: Options,
) {
  removeEventListener(el, EVENT_TYPE.APPEAR, onAppear, options);
}

export function removeDisappearListener(
  el: Element,
  onDisappear: OnEvent,
  options?: Options,
) {
  removeEventListener(el, EVENT_TYPE.DISAPPEAR, onDisappear, options);
}
